syft/workers/base.py

```diff
@@ -911,62 +911,95 @@ def _fetch_protocol_remote(
-                query_item = str(query_item)
-
-                match = False
-                if query_item == str(key):
-                    match = True
-
-                if isinstance(obj, (AbstractObject, FrameworkTensor)):
-                    if obj.tags is not None:
-                        if query_item in obj.tags:
-                            match = True
-
-                    if obj.description is not None:
-                        if query_item in obj.description:
-                            match = True
-
-                if not match:
-                    found_something = False
-
-            if found_something:
-                # set garbage_collect_data to False because if we're searching
-                # for a tensor we don't own, then it's probably someone else's
-                # decision to decide when to delete the tensor.
-                ptr = obj.create_pointer(garbage_collect_data=False, owner=sy.local_worker).wrap()
-                results.append(ptr)
+            if results:
+                results = results.intersection(results_item)
+            else:
+                results = results_item
+
+        if results is not None:
+            return list(results)
+        else:
+            return list()
+
+    def respond_to_search(
+        self, query: Union[List[Union[str, int]], str, int]
+    ) -> List[PointerTensor]:
+        """
+        When remote worker calling search on this worker, forwarding the call and
+        replace found elements by pointers
+        """
+        objects = self.search(query)
+        results = []
+        for obj in objects:
+            # set garbage_collect_data to False because if we're searching
+            # for a tensor we don't own, then it's probably someone else's
+            # decision to decide when to delete the tensor.
+            ptr = obj.create_pointer(
+                garbage_collect_data=False, owner=sy.local_worker, tags=obj.tags
+            ).wrap()
+            results.append(ptr)
 
         return results
 
-    def request_search(self, query: List[str], location: "BaseWorker"):
+    def request_search(self, query: List[str], location: "BaseWorker") -> List:
+        """
+        Add a remote worker to perform a search
+        Args:
+            query: the tags or id used in the search
+            location: the remote worker identity
 
-        result = self.send_msg(SearchMessage(query), location=location)
+        Returns:
+            A list of pointers to the results
+        """
+        results = self.send_msg(SearchMessage(query), location=location)
+        for result in results:
+            self.register_obj(result)
+        return results
 
-        return result
+    def find_or_request(self, tag, location):
+        """
+        Allow efficient retrieval: if the tag is know locally, return the local
+        element. Else, perform a search on location
+        """
+        results = self.find_by_tag(tag)
```

Comment:
Not sure I fully understand for which use case this method exists.
But my remark here is: is it true that if the tag exists, it is not necessarily associated to a pointer to location?

Reply:
The tag can be used to refer to local or remote values. Use case: I download a plan using a tag. Next time I need it I use find_or_request and because I already have it I don't need to re-download it!
Does it answer your question

Reply:
Only the first one :)
But to reformulate my second question: do we want to check that the found pointer really points to location before returning it?
And also, can't it be the case that we have the same tag for pointer to different locations?

Reply:
> do we want to check that the found pointer really points to location before returning it
Excellent point I'll add this check!